applications/openshift/api-server/api_server_admission_control_plugin_AlwaysAdmit/rule.yml

```diff
@@ -29,20 +29,20 @@ ocil_clause: '<tt>AlwaysAdmit</tt> admission control plugin is not set'
 
 ocil: |-
     To verify that the <tt>AlwaysAdmit</tt> admission control plugin is not set, run the following command:
-    <pre>$ oc -n openshift-kube-apiserver get configmap config -o json | jq -r '.data."config.yaml"' | jq '.apiServerArguments."enable-admission-plugins"'</pre>
+    <pre>$ oc -n openshift-kube-apiserver get configmap config -o json | jq -r '.data."config.json"' | jq '.apiServerArguments."enable-admission-plugins"'</pre>
     The output should not contain <tt>AlwaysAdmit</tt>
 
 warnings:
 - general: |-
-    {{{ openshift_cluster_setting("/api/v1/namespaces/openshift-kube-apiserver/configmaps/config") | indent(4) }}}
-
+    {{{ openshift_cluster_setting("{{.var_openshift_kube_apiserver_config}}") | indent(4) }}}
```

Comment:
Wondering do we want to have a global variable like {{{ openshift_cluster_setting("/api/v1/namespaces/{{<hyper-shift-ns>}}/openshift-kube-apiserver/configmaps/config" | indent(4) }}}, instead one variable for the whole api-path?

Reply:
Thank you, it is very important question which relates my initial comment (`how to avoid using two elements in a single yamlfile_value template?`).

There are two API resource path locations in a rule file: one is `.warnings[0]`, another is `.template.vars.filepath_xccdf_variable` (`.filepath_xccdf_variable` is a new element introduced by me).

`.warnings[0]` is used by the API resource collector of Compliance Operator. `parse_arf_result.go` substitutes a placeholder with actual values (e.g., `/api/{{.namespace}}.config{{.configname}}` to `/api/myns/config/myconfig`) before the resource collection, and therefore the API resource collector can read data from the correct resource path (`/api/myns/config/myconfig`).

In contrast, `.template.vars.filepath_xccdf_variable` is used by `oscap` command to read the API resource file. `oscap` command directly reads a data stream XML file, and therefore no substitution can be done before `oscap` opens the API resource file.  It means that we cannot use a description like (`filepath: /api/{{.namespace}}.config{{.configname}}`).

That's why I introduced a new element `.filepath_xccdf_variable` which contains full path for a resource instead of the combination of constant text and variables. This also force me use full path in `.warnings[0]`. 
(If we use notation like `/api/{{.namespace}}.config{{.configname}}` for `.warnings[0]`, then we need to maintain three variables. It just introduces complexity.)

I know that it is not smart, but I can't find any good workaround to avoid introducing the new element.
Do you have any idea to integrate `.template.vars.filepath` and `.template.vars.filepath_xccdf_variable`?

Reply:
Hi @tmishina, I think that's an interesting point! I think there is also another issue, we have some rules that use jq filters, and we use `openshift_filtered_cluster_setting()` for warnings.

Reply:
@Vincent056 
One idea for solving the issue. I'd appreciate it if I can get your comments or other ideas!

---

What we'd like to do is converting an expression including variables like this:

```yaml
filepath: /api/v1/namespaces/{kas_config_namespace}/configmaps/{kas_config_name}
```

into this at the time of datastream build to avoid error at build and test:

```yaml
filepath: |-
	<local_variable>
	  <concat>
	    <variable_component var_ref="ocp_data_root"/>
	    <literal_component>/api/v1/namespaces/</literal_component>
	    <variable_component var_ref="kas_config_namespace"/>
	    <literal_component>/configmaps/</literal_component>
	    <variable_component var_ref="kas_config_name">
	  </concat>
	</local_variable>
```

Let's assume that we generate this value at [preprocess()](https://github.com/ComplianceAsCode/content/blob/master/shared/templates/yamlfile_value/template.py#L4-L31) and store it in `data['filepath_xml']`. In addition, modify [oval.template](https://github.com/ComplianceAsCode/content/blob/master/shared/templates/yamlfile_value/oval.template) as follows.

- use the new data FILEPATH_XML at the [conversion from FILEPATH into XML expression](https://github.com/ComplianceAsCode/content/blob/master/shared/templates/yamlfile_value/oval.template#L21-L30) 
- remain using FILEPATH when the template uses it as an identifier (5 occurrences: [here](https://github.com/ComplianceAsCode/content/blob/master/shared/templates/yamlfile_value/oval.template#L4), [here](https://github.com/ComplianceAsCode/content/blob/master/shared/templates/yamlfile_value/oval.template#L11),  [here](https://github.com/ComplianceAsCode/content/blob/master/shared/templates/yamlfile_value/oval.template#L13),  [here](https://github.com/ComplianceAsCode/content/blob/master/shared/templates/yamlfile_value/oval.template#L34), and [here](https://github.com/ComplianceAsCode/content/blob/master/shared/templates/yamlfile_value/oval.template#L75) ) .

Reply:
I am wondering if we could feed the datastream to go template before we feed it to oscap as datastream, needs some research on it. It will be great if that could work!

Reply:
@Vincent056 I have pushed new commits which rules assume the implementation of the `feed the datastream to go-template before we feed it to oscap as datastream` feature and [the other PR on Compliance Operator](https://github.com/openshift/compliance-operator/pull/750). Please tell me if you have some progress on it, thanks!
Also, now branch has merged the latest contents of `master` branch as of today (2022-01-19).